File: CrackingTheCodingInterview/1-ArraysAndStrings/1.2-checkPermutation.py
# Given two strings, write a method to decide if one is a permutation of the other.

# Initial
# time: O(nlogn), space: O(n)
def checkPermutation1(a, b):
    if len(a) != len(b):
        return False

    a_srtd = "".join(sorted(a)) # O(nlogn)
    b_srtd = "".join(sorted(b)) # O(nlogn)
    for i in range(len(a)):     # O(n)
        if a_srtd[i] != b_srtd[i]:
            return False
    return True

# Optimization
# time: O(n), space: O(n)
def checkPermutation2(a, b):
    if len(a) != len(b):
        return False

    a_hash = {}
    b_hash = {}
    for i in range(len(a)):     # O(n)
        if a[i] in a_hash:
            a_hash[a[i]] += 1
        else:
            a_hash[a[i]] = 0

        if b[i] in b_hash:
            b_hash[b[i]] += 1
        else:
            b_hash[b[i]] = 0

    return a_hash == b_hash     # O(n)

File: CrackingTheCodingInterview/1-ArraysAndStrings/1.2-checkPermutation.js
// Given two strings, write a method to decide if one is a permutation of the other

// Simple
// time: O(nlogn), space: O(n)
function checkPermutation1(a, b) {
	if(a.length !== b.length) {
        return false;
    }
    a_srtd = a.split('').sort().join();		// O(nlogn)
    b_srtd = b.split('').sort().join();
    for (let i = 0; i < a.length; i++) { // O(n)
        if (a_srtd[i] !== b_srtd[i]) {
            return false;
        }
    }
    return true;
}

// Optimization
// time: O(n), space: O(n)
function checkPermutation2(a, b) {
	if(a.length !== b.length) {
        return false;
    }
    a_hash = {}
    b_hash = {}
    for(let i = 0; i < a.length; i++) {
        if(a[i] in a_hash) {
            a_hash[a[i]]++;
        }
        else {
            a_hash[a[i]] = 0;
        }
        if(b[i] in b_hash) {
            b_hash[b[i]]++;
        }
        else {
            b_hash[b[i]] = 0;
        }
    }
    
    // compare objects: O(n)
    let a_keys = Object.keys(a_hash);
    let b_keys = Object.keys(b_hash);
    if(a_keys.length !== b_keys.length) {
        return false;
    }
    for(let key of a_keys) {
        if(a_hash[key] !== b_hash[key]) {
            return false;
        }
    }
    return true;
}